### PR TITLE
fix(frontend): derive `cleaned_by_watermark` using watermark col instead of rhs expression

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

The predicate has been converted into a `StreamNow (+ Project)` and `Exchange`. So we can't check RHS directly, since it will just be an `inputref`.

Add `cleaned_by_watermark` to `StreamDynamicFilter` plan node, so we can test it in planner test.

After this fix, https://github.com/risingwavelabs/risingwave/pull/11572. shows `Delete Ranges to be 12`, and does not issue new delete ranges overtime. 

Need a better e2e test for this behaviour later. Open to suggestions.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs [hunk 1]
```diff
@@ -44,12 +45,42 @@ impl StreamDynamicFilter {
...
+
+    /// 1. Check the comparator.
+    /// 2. RHS input has to be temporal filter.
+    pub fn cleaned_by_watermark(core: &DynamicFilter<PlanRef>) -> bool {
+        let expr = core.predicate();
+        if let Some(ExprImpl::FunctionCall(function_call)) = expr.as_expr_unless_true() {
+            match function_call.func_type() {
+                ExprType::GreaterThan | ExprType::GreaterThanOrEqual => {
+                    let rhs_input = core.right();
+                    if let Some(stream_exchange) = rhs_input.as_stream_exchange() {
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs:68]: Can just check the `rhs_input.watermark()` here

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs:68]: I couldn't find this function (`watermark()`). Could you please elaborate?

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs:68]: Do you mean `!rhs_input.watermark_columns().is_empty()`

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs]: Makes sense to me. We could have other source of watermark.

**Review by chenzl25 — APPROVED** [review]
LGTM

**Comment by BugenZhao** [comment]
> Need a better e2e test for this behaviour later.

Do you mean state cleaning tests? https://github.com/risingwavelabs/risingwave/blob/320d755578505832134d653b918d4e43db442301/src/tests/state_cleaning_test/README.md

File: src/frontend/planner_test/tests/testdata/output/temporal_filter.yaml [hunk 2]
```diff
@@ -47,8 +47,8 @@
     select * from t1 where ts < now() - interval '1 hour' and ts >= now() - interval '2 hour';
   stream_plan: |-
     StreamMaterialize { columns: [ts, t1._row_id(hidden)], stream_key: [t1._row_id], pk_columns: [t1._row_id], pk_conflict: NoCheck }
-    └─StreamDynamicFilter { predicate: (t1.ts < $expr2), output: [t1.ts, t1._row_id] }
-      ├─StreamDynamicFilter { predicate: (t1.ts >= $expr1), output_watermarks: [t1.ts], output: [t1.ts, t1._row_id] }
+    └─StreamDynamicFilter { predicate: (t1.ts < $expr2), output: [t1.ts, t1._row_id], cleaned_by_watermark: false }
```
Comment (BugenZhao) [inline src/frontend/planner_test/tests/testdata/output/temporal_filter.yaml:50]: Can we omit the `false` ones?

**Comment by kwannoel** [comment]
> > Need a better e2e test for this behaviour later.
> 
> Do you mean state cleaning tests? https://github.com/risingwavelabs/risingwave/blob/320d755578505832134d653b918d4e43db442301/src/tests/state_cleaning_test/README.md

Thanks for sharing. Didn't know about that before. 
IIUC state_cleaning_test checks if state is cleaned up. However, in this case I'm looking to test for the presence of delete ranges so I think it needs something slightly different.

To test it I have the following ideas:
- Check the logs to see if they are issued (seems brittle).
- Add Delete Range metrics, test it with an in-memory metrics store.

**Pull request merged**